PYIC-5436 - Tidy up routing for journey action submissions

sammyiyke opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

[PYIC-5436](https://govukverify.atlassian.net/browse/PYIC-5436) - Tidy up routing for journey action submissions

Makes the generic journey action function `handleJourneyAction` robust to handle all `POST` journey avent

Major changes:

- Refactors `handleJourneyAction`
- Removes `handleEscapeAction` and merges it function into `handleJourneyAction`
- Removes `handleEscapeM2b` and merges it function into `handleJourneyAction`
- Removes `handleMultipleDocCheck` and merges it function into `handleJourneyAction`
- Removes `handleUpdateNameDobAction` and merges it function into `handleJourneyAction`

Major tests related changes:

Removes redundant tests for defaulting to `end` when no journey action is provided in form

Reasons:

- Backend validation added in PYIC-5537 removing need to default to `end`
- Previous related functions `handleEscapeAction`, `handleEscapeM2b `,  `handleMultipleDocCheck `, and `handleUpdateNameDobAction` merged into `handleJourneyAction`

<!-- Describe the changes in detail - the "what"-->

### Why did it change

We want to make the generic journey action `handleJourneyAction` robust for all events, so that:

- The form submits the appropriate event - next, f2f etc.

- The handler maps that into a request to the backend POST /journey/next etc.

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5436](https://govukverify.atlassian.net/browse/PYIC-5436)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [ ] No environment variables or secrets were added or changed

<!-- Delete if changes DO NOT include new environment variables or secrets -->
- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks


[PYIC-5436]: https://govukverify.atlassian.net/browse/PYIC-5436?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PYIC-5436]: https://govukverify.atlassian.net/browse/PYIC-5436?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sammyiyke committed:
PYIC-54396 - Initial refactoring for  `handleJourneyAction`

Cleans up the `handleJourneyAction` simplifying it for upcoming changes

sammyiyke committed:
PYIC-5436 - Merged `handleEscapeAction` into  `handleJourneyAction`

Removes the `handleEscapeAction` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Tests --- Merged `handleEscapeAction` into  `handleJourneyAction`

Modifies related tests for --- Removes the `handleEscapeAction` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - changed casing to match similar patterns

sammyiyke committed:
PYIC-5436 - Removed redundant test

Removed this test as it is now redundant:

- Backend validation added in PYIC-5537 removing need to default to `end`
- Previous function `handleEscapeAction` merged into `handleJourneyAction`

sammyiyke committed:
PYIC-5436 Merged `handleEscapeM2b` into `handleJourneyAction` Removes the `handleEscapeM2b` and moved its functionality into the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Test --- Merged `handleEscapeM2b` into `handleJourneyAction`

Modifies related tests for --- Removes the `handleEscapeM2b` and moved its functionality into the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Removed redundant test

Removed this test as it is now redundant:

- Backend validation added in PYIC-5537 removing need to default to `end`
- Previous function `handleEscapeAction` merged into `handleJourneyAction`

sammyiyke committed:
PYIC-5436 - Merged `handleMultipleDocCheck` into `handleJourneyAction`

Removes the `handleMultipleDocCheck` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Test --- Merged `handleMultipleDocCheck` into `handleJourneyAction`

Modifies related tests for --- Removes the `handleMultipleDocCheck` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Removed redundant test

Removed this test as it is now redundant:

- Backend validation added in PYIC-5537 removing need to default to `end`
- Previous function `handleMultipleDocCheck` merged into `handleJourneyAction`

sammyiyke committed:
PYIC-5436 - Merged `handleUpdateNameDobAction` into `handleJourneyAction`

Removes the `handleUpdateNameDobAction` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Tests --- Merged `handleUpdateNameDobAction` into `handleJourneyAction`

Modifies related tests for --- Removes the `handleUpdateNameDobAction` and moved its functionality to the `handleJourneyAction` inline with ticket

sammyiyke committed:
PYIC-5436 - Lint fix

sammyiyke commented on src/app/ipv/middleware.js:
```diff
@@ -358,75 +341,54 @@ module.exports = {
+      "next/bank-account": "bankAccount",
+      "next/passport": "ukPassport",
+      "next/driving-licence": "drivingLicence",
+    };
```
Probably best taking this out of here into a separate file probably in the shared folder. Leaving it here in the meantime until I get feedback on the new implementation

sammyiyke committed:
PYIC-5436 - Changed currentPage to currentPageId

RogVanGDS commented on src/app/ipv/middleware.js:
```diff
@@ -358,87 +341,54 @@ module.exports = {
+      "next/dcmaw": "dcmaw",
+      "next/bank-account": "bankAccount",
+      "next/passport": "ukPassport",
+      "next/driving-licence": "drivingLicence",
```
I was hoping that we might be able to do away with all this mapping, given we are now validating journey events on the backend, and posting some invalid journey events isn't as risky as it was before

I expect we will still need a list of pages that need to default to back instead of next, but we should be able to either change the radio button forms so that the journey event is already in the right format, or update the journey map to match if we need to

sammyiyke replied:
Good shout. I do agree, but was optimistic about changing it earlier.

Joe-Edwards-GDS commented on src/app/ipv/middleware.js:
```diff
@@ -358,87 +341,54 @@ module.exports = {
+        checkForSessionId(req, res);
       } else {
-        await handleJourneyResponse(req, res, "end", currentPageId);
+        checkForIpvAndOauthSessionId(req, res);
```
Do you know what this distinction is based on? I can't see any obvious basis for the difference.

I'm wondering if we can just use `checkForIpvAndOauthSessionId` for everything, and then you don't need the list of pages to distinguish :)

sammyiyke replied:
I didn't look that deep into it. But on the surface I figure that some pages **require** only the ipvSession ID while some **require** both the ipvSession and the OauthSession ID. And they fail if any of the required session ID is not present. 

I am not exactly sure why, as I think by default journeys should have both IDs present at each stage. 

I would look deeper and see what the reasoning behind the differentiation is and if changing it would break any test

Joe-Edwards-GDS replied:
I think this method looks for the `ipvSessionId` OR `clientOauthSessionId` (not AND), so I was hoping we could just use that for everything.

It's fine to leave, since you've not changed the logic, but I suspect it isn't quite right.

Joe-Edwards-GDS commented on src/app/ipv/middleware.js:
```diff
@@ -358,87 +341,54 @@ module.exports = {
+      if (action === "build-client-oauth-response") {
+        req.session.ipAddress = req?.session?.ipAddress
+          ? req.session.ipAddress
+          : getIpAddress(req);
```
This seems suspicious, but I'm scared of changing anything in analytics without fully understanding it! (I'm a bit surprised we don't just set the IP address on every request independently...)

Leave this in for now :)

Joe-Edwards-GDS commented on src/app/ipv/middleware.js:
```diff
@@ -358,87 +341,54 @@ module.exports = {
     }
   },
+
+  // make considerations for handleEscapeAction pages with no req.body?.journey that defaults to `end`
```
Can we delete this comment?

Joe-Edwards-GDS commented on src/app/ipv/middleware.js:
```diff
@@ -358,87 +341,54 @@ module.exports = {
+      "next/dcmaw": "dcmaw",
+      "next/bank-account": "bankAccount",
+      "next/passport": "ukPassport",
+      "next/driving-licence": "drivingLicence",
```
Do we need this mapping? Can we instead update the page templates to pass the correct event (e.g. updating `pyi-escape-m2b-form.njk` to just have `bankAccount`)

Joe-Edwards-GDS reviewed (commented):
A couple of comments but I think this is looking much neater!

sammyiyke committed:
PYIC-5669 - PR Review changes

Simplifies and removes mapping for submitted form journey values

- Fixes the forms to return the correct expected value
- Use returned value in form submission
